src/rbac/default_role_manager.rs

```diff
@@ -0,0 +1,353 @@
+            panic!("error domain length");
+        }
+        if !self.has_role(&name1) || !self.has_role(&name2) {
+            panic!("name12 error");
+        }
+        let role1 = self.create_role(&name1);
+        let role2 = self.create_role(&name2);
+        role1.borrow_mut().delete_role(Rc::clone(&role2));
+    }
+
+    fn has_link(&mut self, name1: &str, name2: &str, domain: Option<&str>) -> bool {
+        let mut name1 = name1.to_owned();
+        let mut name2 = name2.to_owned();
+        if let Some(domain_val) = domain {
+            name1 = format!("{}::{}", domain_val, name1);
+            name2 = format!("{}::{}", domain_val, name2);
+        }
+        if name1 == name2 {
+            return true;
+        }
+        if !self.has_role(&name1) || !self.has_role(&name2) {
+            return false;
+        }
+        let role1 = self.create_role(&name1);
+        return role1.borrow().has_role(&name2, self.max_hierarchy_level);
+    }
+
+    fn get_roles(&mut self, name: &str, domain: Option<&str>) -> Vec<String> {
+        let mut name = name.to_owned();
+        if let Some(domain_val) = domain {
+            name = format!("{}::{}", domain_val, name);
+        }
+        if !self.has_role(&name) {
+            return vec![];
+        }
+        let role = self.create_role(&name);
+
+        if let Some(domain_val) = domain {
+            return role
+                .borrow()
+                .get_roles()
+                .iter()
+                .map(|x| x[domain_val.len() + 2..].to_string())
+                .collect();
+        } else {
+            return role.borrow().get_roles();
+        }
+    }
+
+    fn get_users(&self, name: &str, domain: Option<&str>) -> Vec<String> {
+        let mut name = name.to_owned();
+        if let Some(domain_val) = domain {
+            name = format!("{}::{}", domain_val, name);
+        }
+        if !self.has_role(&name) {
+            return vec![];
+        }
+
+        let mut names: Vec<String> = vec![];
+        for (_key, role) in self.all_roles.iter() {
+            if role.borrow().has_direct_role(&name) {
+                names.push(role.borrow().name.clone());
+            }
+        }
+        if let Some(domain_val) = domain {
+            return names
+                .iter()
+                .map(|x| x[domain_val.len() + 2..].to_string())
+                .collect();
+        }
+        return names;
+    }
+
+    fn print_roles(&self) {
+        println!("current role manager roles: {:?}", self.all_roles.clone());
+    }
+
+    fn clear(&mut self) {
+        self.all_roles = HashMap::new();
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Role {
+    name: String,
+    roles: Vec<Rc<RefCell<Role>>>,
+}
+
+impl Role {
+    fn new(name: String) -> Self {
+        return Role {
+            name,
+            roles: vec![],
+        };
+    }
+
+    fn add_role(&mut self, other_role: Rc<RefCell<Role>>) {
+        for role in self.roles.iter() {
```

Comment:
The explicit call to `iter` is probably not necessary. Clippy, a linter built with rustc, would tell you this.

Reply:
This whole loop can be implemented with `iter().any(|_| ...)`.

Reply:
`iter` can not be omitted here, we are passing a `&mut self` here, removing `iter` will cause a `move of mutable reference error`.

You are right, it better to use `iter().any` here, will fix later.

Reply:
fixed in https://github.com/casbin/casbin-rs/pull/20/commits/bc1c6e26981fb4708d40b5822dc4d506a4d4aead